File: 프로그래머스/2/176962. 과제 진행하기/과제 진행하기.py
import sys

def to_min(s):
    h, m = map(int, s.split(":"))
    return h * 60 + m

def solution(plans):
    answer = []
    waiting_plans = []
    plans = [(name, to_min(start), int(playtime)) for name, start, playtime in plans]
    plans.sort(key=lambda x: x[1])

    for i in range(len(plans)):
        name, start, playtime = plans[i]

        if i + 1 < len(plans):
            available = plans[i+1][1] - start

            if playtime <= available:
                answer.append(name)
                remaining = available - playtime
                while waiting_plans and remaining > 0:
                    w_name, w_time = waiting_plans.pop()
                    if w_time <= remaining:
                        answer.append(w_name)
                        remaining -= w_time
                    else:
                        waiting_plans.append((w_name, w_time - remaining))
                        break
            else:
                waiting_plans.append((name, playtime - available))
        else:
            answer.append(name)
            while waiting_plans:
                answer.append(waiting_plans.pop()[0])
    return answer

File: archive/programmers/176962/과제 진행하기.js
function solution(plans) {
    var answer = [];
   
    
    plans.forEach((plan) => {
        plan[1] = changeHourToMinute(plan[1]);
        plan[2] = +plan[2];
    })
    plans.sort((a,b) => a[1]- b[1]);
    
    
    const stack = [];
  
    for (let i = 0; i < plans.length - 1; i++) {
        const [name, start, playtime] = plans[i];
        
        if (start + playtime <= plans[i + 1][1]) {
            answer.push(name);
        
            let availableTime = plans[i + 1][1] - start - playtime;
            
            while (stack.length) {
                const [currentSubject, currentTime] = stack.pop();
                if (currentTime <= availableTime) {
                    availableTime -= currentTime;
                    answer.push(currentSubject);
                } else {
                    stack.push([currentSubject, currentTime - availableTime]);
                    break;
                }
            }
        } else {
            stack.push([name, playtime - (plans[i + 1][1] - start)]);
        }
    }
    answer.push(plans[plans.length - 1][0]);
    
    while (stack.length) {
        answer.push(stack.pop()[0]);
    }
    
    return answer;

    
    function changeHourToMinute(prev) {
    const [hour, minute] = prev.split(":");
    return +hour * 60 + +minute;
}
}

